noggin/tests/unit/controller/test_authentication.py

```diff
@@ -91,6 +91,27 @@ def test_login(client, dummy_user):
     assert session.get("noggin_session") is not None
 
 
+@pytest.mark.vcr()
+def test_login_with_otp(client, dummy_user):
+    """Test a successful Login with password + otp"""
+    result = client.post(
+        '/',
+        data={
+            "login-username": "dummy",
+            "login-password": "dummy_password",
+            "login-otp": "123456",
```

Comment:
Hmm, this should not work, because the dummy user does not have an OTP token enabled by default in tests. You have to use the `dummy_user_with_otp` fixture instead of the `dummy_user` fixture. You can look at the tests in the  `test_user_otp.py` module to see how you can get the current value of the token for your otp form field.

I suppose you copied the cassette and edited the content but that's not how cassettes are generated, it's much simpler: just run the unit tests inside the vagrant VM and the cassette will be automatically created.

Reply:
Yes, I did not how VCR works. I've searched about it and after discovering the `--vcr-record` flag I finally was able to create news cassettes 😊 

I have some doubts related to how to make a not logged user with an OTP token to test it because after so many tries I failed to create a new `dummy_user`. The current `dummy_user_with_otp` is already logged and doesn't fit to test the login. I created in a new fixture a user (I refactored just a little bit the `make_user` function) and I added a new otp token after setting the password with `changed_password`: 

```
@pytest.fixture
def make_user(ipa_testing_config, app):
    created_users = []

    def _make_user(name, with_otp=False):
        now = datetime.datetime.utcnow().replace(microsecond=0)
        password = f'{name}_password'
        ipa_admin.user_add(
            a_uid=name,
            o_givenname=name.title(),
            o_sn='User',
            o_cn=f'{name.title()} User',
            o_mail=f"{name}@example.com",
            o_userpassword=password,
            o_loginshell='/bin/bash',
            fascreationtime=f"{now.isoformat()}Z",
        )
        created_users.append(name)
        ipa = untouched_ipa_client(app)
        ipa.change_password(name, new_password=password, old_password=password)

        if with_otp:
            otp_token = ipa_admin.otptoken_add(o_ipatokenowner=name, o_description=f"{name}'s token")
            otp_token = OTPToken(otp_token["result"])
            return otp_token

    yield _make_user

    for name in created_users:
        ipa_admin.user_del(name)
```

When I test the login sending the OTP code (just like the following example) I get `invalid-password` error from the login and (of course) the test fails.

```
result = client.post(
        '/',
        data={
            "login-username": "dummy",
            "login-password": "dummy_password",
            "login-otp": otp,
            "login-submit": "1",
        },
        follow_redirects=True,
    )
```
Could you please tell me if the connection to the IPA with the token should be mocked?

Reply:
> The current `dummy_user_with_otp` is already logged and doesn't fit to test the login

You're right! So what I think you should do is either:
- create a fixture that is similar to what `dummy_user_with_otp` does except it takes a `dummy_user` as dependency (the function argument) instead of a `logged_in_dummy_user`, and adds the token with `ipa_admin` instead of `ipa`
- or add the token inside your test function, but use the `cleanup_dummy_tokens` fixture to remove it afterwards

I recommend doing composition with fixtures (new fixtures which depend on other fixtures) rather than refactoring existing fixtures to add more features.

Then, in your test, you can do something like `current_otp = get_otp(otp_secret_from_uri(token.uri))` to get the current value of the token and add it to your POST.

Feel free to ask me questions if you're blocked.

Reply:
Hello again! I'm still struggling to finish successfully the test. I did the fixture just like you said but when I run the test, in fact when I run the tests recording the cassette in the VM, the ones which use a user with tokens added will fail. All the time these tests return [`kinit: Preauthentication failed while getting initial credentials`](https://lists.fedorahosted.org/archives/list/freeipa-users@lists.fedorahosted.org/message/XT5GZFGHVEQH2LH56UYC56EIXX2N6PTH/). 

This image shows an example of a failure. Instead of return the user profile page, it is the login page with the message of `bad credentials` added to password field in the login form. 
![image](https://user-images.githubusercontent.com/9274112/115335918-2721d280-a16c-11eb-8195-57947f97a9f6.png)

I updated the PR with my current code of the tests and not the cassettes. I'm quite sure the error comes from my freeipa local VM, could you please can help me to configure the freeipa VM to avoid this error? I'm sorry to bother you with this issue but I couldn't solve it by myself.

Reply:
My mistake! We made a change in the hash algorithm used in OTP tokens a while ago and forgot to update the test functions. Please rebase on the HEAD of the dev branch, it should make this error disappear. Sorry about that!
And thanks for your perseverance.